Protobuf extension

knoguchi opened the pull request:
I've created protobuf-extensions per #3508 and #3509.

1. Factor out ProtoBuf parser from `processing` module to `extensions-contrib/protobuf-extensions`
2. Migrate ProtoBuf from v2 to v3
3. Add flattenSpec support.
4. Support composite (nested) message types.  Original implementation had an assumption that the descriptor file contains one message type only.  [See this line](https://github.com/druid-io/druid/blob/master/processing/src/main/java/io/druid/data/input/ProtoBufInputRowParser.java#L120).  In order to specify the message type, I added a mandatory key `protoMessageType`.
5. Load the descriptor file from a URL as well as classpath.  Original implementation expected the descriptor file in the classpath.   It was inconvenient when using with the Kafka indexing service that can create indexing jobs dynamically.

The com.google.protobuf package might need shading because some part of Druid is still using protobuf 2.5.0.  I will include the shade plugin in pom.xml if necessary.

[Here is an example JSON](https://gist.github.com/knoguchi/e0f5fd654ab52158e257d993287a4d63) that works with the Kafka indexing service.

gianm commented:
@knoguchi Shading would be good, I think, since we've had problems with protobuf version conflicts in the past -- in particular with Hadoop.

Also, are you using this in production?

gianm commented on extensions-contrib/protobuf-extensions/src/main/java/io/druid/data/input/protobuf/ProtoBufExtensionsModule.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class ProtoBufExtensionsModule implements DruidModule
```
Is ProtoBuf really right? I usually see it styled as "Protobuf" with a lowercase b.

knoguchi replied:
Indeed Protobuf is the correct case.  I followed the naming convention of the old code to minimize potential breakages.

gianm commented on extensions-contrib/protobuf-extensions/src/main/java/io/druid/data/input/protobuf/ProtoBufInputRowParser.java:
```diff
@@ -0,0 +1,138 @@
+import java.util.Map;
+
+@JsonTypeName("protobuf")
+public class ProtoBufInputRowParser implements ByteBufferInputRowParser {
```
Don't include commented out code.

knoguchi replied:
Will do.

gianm commented on extensions-contrib/protobuf-extensions/src/main/java/io/druid/data/input/protobuf/ProtoBufInputRowParser.java:
```diff
@@ -0,0 +1,138 @@
+import java.nio.ByteBuffer;
+import java.util.Map;
+
+@JsonTypeName("protobuf")
```
I think you don't need both this and registration of a NamedType in the Guice module, although I'm not totally sure.

knoguchi replied:
It's from the old code.  I will remove it up on test.

gianm commented on extensions-contrib/protobuf-extensions/src/main/java/io/druid/data/input/protobuf/ProtoBufInputRowParser.java:
```diff
@@ -0,0 +1,138 @@
+    try {
+      DynamicMessage message = DynamicMessage.parseFrom(descriptor, ByteString.copyFrom(input));
+      json = JsonFormat.printer().print(message);
+      // log.trace(json);
```
Don't include commented out code

knoguchi replied:
Will do

gianm commented on extensions-contrib/protobuf-extensions/src/main/java/io/druid/data/input/protobuf/ProtoBufInputRowParser.java:
```diff
@@ -0,0 +1,138 @@
+      throw new ParseException(e, "Cannot read " + descriptorFilePath);
+    }
+
+    Descriptor desc = dynamicSchema.getMessageDescriptor(protoMessageType);
```
It looks like the old code would always take the first message type, and you're adding a protoMessageType to control which one is read. For compatibility with the old code, this should read the first message type if protoMessageType is null.

knoguchi replied:
I will allow protoMessageType null and read the first one only when the number of the descriptors in the descriptor file is 1.

I thought about the backward compatibility.  If we simply pick the first descriptor it would cause non deterministic problems in the future in unexpected way because the order of the descriptors in the descriptor files is not controllable by us.  One day when you recompile the descriptor file with one or more .proto files, the parser may (may not) pick unintended descriptor and stop working.

gianm commented on extensions-contrib/protobuf-extensions/src/main/java/io/druid/data/input/protobuf/ProtoBufInputRowParser.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    DynamicSchema dynamicSchema = null;
+    try {
```
The old code did `DescriptorProtos.FileDescriptorSet.parseFrom(fin)` and this new code instead uses DynamicSchema from a new protobuf-dynamic library. What's the rationale for this change? Is it compatible with descriptor files people might have been using with the old code?

knoguchi replied:
The descriptor files are what the `protoc` compiler generates.  Both old code and my code use the same descriptor files.   The rationale is the old code does not work for nested message types.   The protobuf-dynamic takes care of loading descriptors for the nested message types.   You only have to specify the top level message type which is provided by the `protoMessageType`.

gianm commented on extensions-contrib/protobuf-extensions/src/main/java/io/druid/data/input/protobuf/ProtoBufInputRowParser.java:
```diff
@@ -0,0 +1,138 @@
+      throw new ParseException(e, "Protobuf message could not be parsed");
+    }
+
+    Map<String, Object> record = parser.parse(json);
```
I guess this means that the parseSpec has to be a "json" parseSpec, since we're parsing protobuf by turning it into json first. This isn't great for performance, but the old code wasn't a paragon of efficiency either, so I'm willing to let it slide.

Although: have you looked into whether it'd be possible to apply the flattenSpec to the DynamicMessage directly? If doable, that should be faster than either this code or the old code.

knoguchi replied:
I'm well aware of the performance issue.  What I looked at is the [JsonProvider](https://github.com/jayway/JsonPath/blob/master/json-path/src/main/java/com/jayway/jsonpath/spi/json/JsonProvider.java) and [MappingProvider](https://github.com/jayway/JsonPath/blob/master/json-path/src/main/java/com/jayway/jsonpath/spi/mapper/MappingProvider.java) SPI in the JSONPath module.  The benefit is we could provide the `flattenSpec` not only for Protobuf but potentially any format e.g. AvroProvider as I discussed at #3505.

Unfortunately the provider interface requires [String](https://github.com/jayway/JsonPath/blob/master/json-path/src/main/java/com/jayway/jsonpath/spi/json/JsonProvider.java#L32) or [InputStream](https://github.com/jayway/JsonPath/blob/master/json-path/src/main/java/com/jayway/jsonpath/spi/json/JsonProvider.java#L41). 

I actually have  ProtobufProvider and ProtobufMapper implementation that takes DynamicMessage via InputStream.  But it's still not zero-copy and hacky.  Therefore I did not include them in this PR.

If we modify the JSONPath that takes `Object` type, we can implement zero copy provider.  And if we do, probably we have to maintain the fork.  I don't think JSONPath author would accept some random object provider that is not JSON ;-)

fjy commented on extensions-contrib/protobuf-extensions/pom.xml:
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
missing license header

fjy commented:
Took a quick look. Sames issues as what @gianm pointed out but generally this looks good to me.

fjy commented:
👍

knoguchi commented:
I don't know why Travis failed for the last commit.  I just let it re-run, and it looks ok this time.  I think I addressed all the points.   Remaining issue is the performance but that will be another PR.

There is a PR #2354 that seems to perform better.   I might reuse some code from the PR and create the JSONPath Protobuf provider.

knoguchi commented:
Travis is having timeout or dead lock problem.  The test passes in my local machine.

fjy commented:
@knoguchi can you fix up the conflicts and we'll get this merged before next release

knoguchi commented:
ProtoTestEventWrapper.java is a generated class by the protobuf compiler.   It seems the file was further modified by some kind of tool (Error Prone?) in the master branch besides the license header.  IMO we shouldn't edit the generated code unless absolutely necessary.  To fix the conflict, I have to apply the same tool, but I don't know how.

gianm commented:
@knoguchi I think it'd be fine to re-generate the file using the generator, and add exclusions for it to the errorprone config. For example see common/pom.xml where antlr-generated files are excluded from certain checks. The config looks like this:

```
                    <plugin>
                        <groupId>org.apache.maven.plugins</groupId>
                        <artifactId>maven-compiler-plugin</artifactId>
                        <configuration>
                            <compilerArgs>
                                <!-- Antlr-generated classes miss @Override, that is not easy to fix -->
                                <arg>-Xep:MissingOverride:WARN</arg>
                            </compilerArgs>
                        </configuration>
                    </plugin>
```

fjy commented:
@knoguchi can you fix merge conflicts and I'll merge this

knoguchi commented:
@fjy the conflict has been resolved.

jihoonson commented:
> Migrate ProtoBuf from v2 to v3

Hi @knoguchi, I wonder how the version migration is done in this patch because it looks like the dependency is actually updated in https://github.com/druid-io/druid/pull/3682. Some users want to know which druid version starts supporting protobuf v3 (https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/druid-user/DaBeEA8Ca50/PXS7l4nUBQAJ).

knoguchi commented:
Looks like Apache Calcite Avatica introduced by #3682 requires Protobuf3.  I started working on this PR last year Sept or Oct before SQL module was introduced.   Either way, the Protobuf3 com.google.protobuf package is shaded in this extension because Hadoop libraries still depend on Protobuf 2.  Maybe it's no longer necessary.

Out of curiosity I grep'd protobuf-java.  hdfs-storage acually "excluded" protobuf-java.  

```
$ git grep -A2 protobuf-java
extensions-contrib/orc-extensions/pom.xml:                    <artifactId>protobuf-java</artifactId>
extensions-contrib/orc-extensions/pom.xml-                </exclusion>
extensions-contrib/orc-extensions/pom.xml-                <exclusion>
--
extensions-core/druid-kerberos/pom.xml:          <artifactId>protobuf-java</artifactId>
extensions-core/druid-kerberos/pom.xml-        </exclusion>
extensions-core/druid-kerberos/pom.xml-        <exclusion>
--
extensions-core/hdfs-storage/pom.xml:                <artifactId>protobuf-java</artifactId>
extensions-core/hdfs-storage/pom.xml-              </exclusion>
extensions-core/hdfs-storage/pom.xml-              <exclusion>
--
extensions-core/protobuf-extensions/pom.xml:      <artifactId>protobuf-java</artifactId>
extensions-core/protobuf-extensions/pom.xml-      <version>${protobuf.version}</version>
extensions-core/protobuf-extensions/pom.xml-    </dependency>
--
extensions-core/protobuf-extensions/pom.xml:      <artifactId>protobuf-java-util</artifactId>
extensions-core/protobuf-extensions/pom.xml-      <version>${protobuf.version}</version>
extensions-core/protobuf-extensions/pom.xml-    </dependency>
--
pom.xml:                <artifactId>protobuf-java</artifactId>
pom.xml-                <version>3.1.0</version>
pom.xml-            </dependency>
```

jihoonson commented:
Ah, maybe there was a version conflict problem before this patch. Cool. Thank you for the explanation.
